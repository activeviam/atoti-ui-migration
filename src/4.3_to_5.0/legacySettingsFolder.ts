import type { ContentRecord } from "@activeviam/activeui-sdk";

const defaultPreferences = {
  allow: [],
  deny: [],
  map: {
    "global.theme": "light-activeviam",
    "global.locale": "en-US",
  },
};

const user1Preferences = {
  allow: [],
  deny: [],
  map: {
    "userFilters.enabled": true,
  },
};

const user1Permissions = {
  map: {
    "userFilters.enabled": false,
  },
};

const roleUserPreferences = {
  allow: [],
  deny: [],
  map: {
    "tree.search.maxResults": 7,
  },
};

const adminPreferences = {
  map: {
    "tree.search.maxResults": 10,
    "global.theme": "dark-activeviam",
    "userFilters.enabled": true,
    showLegacyCharts: true,
    showLegacyMaps: true,
    defaultPermissions: {
      owners: ["admin"],
      readers: ["admin"],
    },
    "project.content-editor.configuration": {
      autoSwitchToFieldsOnEmptyWidget: true,
      advancedModeDropdownHidden: false,
      hideFieldsControls: false,
      "mdx-common": {
        regularModeTab: "filters",
        advancedModeEnabled: false,
        advancedModeTab: "mdx",
      },
    },
    "user.contextValues": {
      EquityDerivativesCube: {},
      EquityDerivativesCubeDist: {},
    },
    "user.filters": {
      EquityDerivativesCube: [],
      EquityDerivativesCubeDist: [
        "[Geography].[City].[ALL].[AllMember].[Berlin]",
      ],
    },
    "filtering.hierarchyToFilterType.[Geography].[City]": "explicit",
    "filtering.hierarchyToFilterType.[Currency].[Currency]": "explicit",
    "filtering.hierarchyToFilterType.[Trades].[Trades]": "topcount",
    "servers.alias": {
      "https://activepivot-ranch.activeviam.com:5900": "http://localhost:8080",
    },
  },
};

/**
 * The content of a legacy /ui/settings folder on a Content Server, useful for unit tests.
 */
export const legacySettingsFolder: ContentRecord = {
  entry: {
    isDirectory: true,
    owners: ["ROLE_CS_ROOT"],
    readers: ["ROLE_USER"],
    timestamp: 1607879735685,
    lastEditor: "admin",
    canRead: true,
    canWrite: true,
  },
  children: {
    default: {
      entry: {
        isDirectory: true,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_USER"],
        timestamp: 1607879735685,
        lastEditor: "admin",
        canRead: true,
        canWrite: true,
      },
      children: {
        preferences: {
          entry: {
            content: JSON.stringify(defaultPreferences),
            isDirectory: false,
            owners: ["ROLE_CS_ROOT"],
            readers: ["ROLE_USER"],
            timestamp: 1607879735685,
            lastEditor: "admin",
            canRead: true,
            canWrite: true,
          },
        },
      },
    },
    roles: {
      entry: {
        isDirectory: true,
        owners: ["ROLE_CS_ROOT"],
        readers: ["ROLE_USER"],
        timestamp: 1607879735685,
        lastEditor: "admin",
        canRead: true,
        canWrite: true,
      },
      children: {
        ROLE_USER: {
          entry: {
            isDirectory: true,
            owners: ["ROLE_CS_ROOT"],
            readers: ["ROLE_USER"],
            timestamp: 1607879735685,
            lastEditor: "admin",
            canRead: true,
            canWrite: true,
          },
          children: {
            preferences: {
              entry: {
                owners: ["ROLE_CS_ROOT"],
                readers: ["ROLE_USER"],
                timestamp: 1607879735685,
                lastEditor: "admin",
                canRead: true,
                canWrite: true,
                content: JSON.stringify(roleUserPreferences),
              },
            },
          },
        },
      },
    },
    users: {
      entry: {
        isDirectory: true,
        owners: ["ROLE_USER"],
        readers: ["ROLE_USER"],
        timestamp: 1607879735685,
        lastEditor: "admin",
        canRead: true,
        canWrite: true,
      },
      children: {
        user1: {
          entry: {
            isDirectory: true,
            owners: ["user1"],
            readers: ["user1"],
            timestamp: 1607879735685,
            lastEditor: "admin",
            canRead: true,
            canWrite: true,
          },
          children: {
            preferences: {
              entry: {
                content: JSON.stringify(user1Preferences),
                isDirectory: false,
                owners: ["user1"],
                readers: ["user1"],
                timestamp: 1607879735685,
                lastEditor: "admin",
                canRead: true,
                canWrite: true,
              },
            },
            permissions: {
              entry: {
                content: JSON.stringify(user1Permissions),
                isDirectory: false,
                owners: ["ROLE_CS_ROOT"],
                readers: ["user1"],
                timestamp: 1607879735685,
                lastEditor: "admin",
                canRead: true,
                canWrite: true,
              },
            },
          },
        },
        admin: {
          entry: {
            isDirectory: true,
            owners: ["admin"],
            readers: ["admin"],
            timestamp: 1607879735685,
            lastEditor: "admin",
            canRead: true,
            canWrite: true,
          },
          children: {
            preferences: {
              entry: {
                content: JSON.stringify(adminPreferences),
                isDirectory: false,
                owners: ["admin"],
                readers: ["admin"],
                timestamp: 1607879735685,
                lastEditor: "admin",
                canRead: true,
                canWrite: true,
              },
            },
          },
        },
      },
    },
  },
};
