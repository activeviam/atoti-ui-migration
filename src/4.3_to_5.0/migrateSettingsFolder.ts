import _cloneDeep from "lodash/cloneDeep";
import _mapValues from "lodash/mapValues";
import _uniq from "lodash/uniq";
import _pick from "lodash/pick";
import _reduce from "lodash/reduce";
import {
  Activity,
  ContentRecord,
  Settings,
  MdxString,
  unquote,
} from "@activeviam/activeui-sdk-5.0";
import { emptyUIFolder } from "@activeviam/content-server-initialization-5.0";

const emptySettingsFolders = _pick(
  emptyUIFolder.children,
  "organization_settings",
  "users",
);

/**
 * Returns the converted settings map, ready to be used by ActiveUI 5.
 */
function migrateSettingsMap(legacySettingsMap: {
  [settingKey: string]: any;
}): Partial<Settings> {
  const migratedSettingsMap: Partial<Settings> = {};

  const theme = legacySettingsMap["global.theme"];
  if (["light-activeviam", "dark-activeviam"].includes(theme)) {
    migratedSettingsMap.theme = theme;
  }

  const searchMaxResults = legacySettingsMap["tree.search.maxResults"];
  if (searchMaxResults !== undefined) {
    migratedSettingsMap["search.maxResults"] = searchMaxResults;
  }

  const areUserFiltersEnabled = legacySettingsMap["userFilters.enabled"];
  if (areUserFiltersEnabled !== undefined) {
    migratedSettingsMap["userFilters.areEnabled"] = areUserFiltersEnabled;
  }

  const legacyDrillthroughColumns:
    | {
        [serverUrl: string]: {
          [cubeName: string]: {
            functionName: "Value" | "Caption" | "MemberValue" | "MemberCaption";
            columnName: string;
          }[];
        };
      }
    | undefined =
    legacySettingsMap["widgets.Tabular.drillthrough.selectedColumns"];

  if (legacyDrillthroughColumns !== undefined) {
    const migratedDrillthroughColumns: Settings["drillthrough.defaultSelectedColumns"] =
      {};
    // The legacy settings are per server and per cube.
    // The new ones are only per cube: the assumption is that no customer has multiple servers containing a cube with the same name but exposing a different data model.
    for (const serverUrl in legacyDrillthroughColumns) {
      for (const cubeName in legacyDrillthroughColumns[serverUrl]) {
        if (migratedDrillthroughColumns[cubeName] === undefined) {
          migratedDrillthroughColumns[cubeName] = [];
        }
        migratedDrillthroughColumns[cubeName].push(
          ...legacyDrillthroughColumns[serverUrl][cubeName].map(
            ({ columnName }) => columnName,
          ),
        );
      }
    }

    // Remove duplicates.
    migratedSettingsMap["drillthrough.defaultSelectedColumns"] = _mapValues(
      migratedDrillthroughColumns,
      _uniq,
    );
  }

  /* Smart filtering settings start. */
  const isSmartFilteringDisabled =
    legacySettingsMap["memberSelection.smartFiltering"] === false;
  if (isSmartFilteringDisabled) {
    migratedSettingsMap["smartFiltering.isEnabled"] = false;
  } else {
    const nonEmptyEvaluationMeasureUniqueName =
      legacySettingsMap["memberSelection.smartFiltering.nonEmptyMeasure"];
    if (nonEmptyEvaluationMeasureUniqueName !== undefined) {
      const [, nonEmptyEvaluationMeasureName] = unquote(
        nonEmptyEvaluationMeasureUniqueName,
      );
      // @ts-expect-error TypeScript rightfully complains that this setting key does not exist in 5.0
      // It was added only in 5.1 indeed, but it makes sense to migrate it as an intermediate step.
      migratedSettingsMap["smartFiltering.nonEmptyEvaluationMeasureName"] =
        nonEmptyEvaluationMeasureName;
    }

    const hierarchiesToIgnoreWhenSmartFiltering = _reduce(
      legacySettingsMap,
      (hierarchyUniqueNames: string[], settingValue, settingKey) => {
        const [, hierarchyUniqueName] = settingKey.split(
          "memberSelection.smartFiltering.ignoreFromContext.",
        );
        if (hierarchyUniqueName && settingValue === true) {
          hierarchyUniqueNames.push(hierarchyUniqueName);
        }
        return hierarchyUniqueNames;
      },
      [],
    );
    if (hierarchiesToIgnoreWhenSmartFiltering.length > 0) {
      migratedSettingsMap["smartFiltering.ignoredHierarchies"] =
        hierarchiesToIgnoreWhenSmartFiltering;
    }
  }
  /* Smart filtering settings end. */

  return migratedSettingsMap;
}

/**
 * Returns the legacy settings map corresponding to the given legacy settings subfolder,
 * where the settings from the "permissions" file take precedence over those from the "preferences" file.
 */
function getLegacySettingsMap(legacySettingsSubfolder: ContentRecord): {
  [settingKey: string]: any;
} {
  const { permissions: permissionsFile, preferences: preferencesFile } =
    legacySettingsSubfolder?.children || {};
  const [permissions, preferences] = [permissionsFile, preferencesFile].map(
    (file) => (file ? JSON.parse(file.entry.content).map : {}),
  );
  return { ...preferences, ...permissions };
}

/**
 * Returns the folders corresponding to the converted ActiveUI 4 settings folder, ready to be used by ActiveUI 5.
 */
export function migrateSettingsFolder(legacySettingsFolder?: ContentRecord): {
  organization_settings: ContentRecord;
  users: ContentRecord;
} {
  const migratedSettingsFolders: {
    organization_settings: ContentRecord;
    users: ContentRecord;
  } = _cloneDeep(emptySettingsFolders);

  const { default: legacyDefaultFolder, users: legacyUsersFolder } =
    legacySettingsFolder?.children ?? {};

  if (legacyDefaultFolder) {
    const legacyDefaultSettingsMap = getLegacySettingsMap(legacyDefaultFolder);
    const migratedDefaultSettingsMap = migrateSettingsMap(
      legacyDefaultSettingsMap,
    );
    migratedSettingsFolders.organization_settings.entry = {
      content: JSON.stringify(migratedDefaultSettingsMap),
      owners: ["ROLE_CS_ROOT"],
      readers: ["ROLE_USER"],
    };
  }

  if (legacyUsersFolder?.children) {
    migratedSettingsFolders.users.children = _mapValues(
      legacyUsersFolder.children,
      (userFolder, userName) => {
        const legacyUserSettingsMap = getLegacySettingsMap(userFolder);
        const migratedUserSettingsMap = migrateSettingsMap(
          legacyUserSettingsMap,
        );

        const activity: Partial<Activity<"serialized">> = {};
        const legacyUserFilters: MdxString[] =
          legacyUserSettingsMap["user.filters"];
        if (legacyUserFilters !== undefined) {
          activity.userFilters = Object.values(legacyUserFilters).flat();
        }

        const migratedUserFolder = {
          entry: userFolder.entry,
          children: {
            settings: {
              entry: {
                content: JSON.stringify(migratedUserSettingsMap),
                owners: [userName],
                readers: [userName],
              },
            },
            activity: {
              entry: {
                content: JSON.stringify(activity),
                owners: [userName],
                readers: [userName],
              },
            },
          },
        };
        return migratedUserFolder;
      },
    );
  }

  return migratedSettingsFolders;
}
