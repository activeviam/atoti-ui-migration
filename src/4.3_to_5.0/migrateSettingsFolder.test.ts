import { legacySettingsFolder } from "./legacySettingsFolder";
import { migrateSettingsFolder } from "./migrateSettingsFolder";

describe("migrateSettingsFolder", () => {
  it("returns the folders corresponding to the converted ActiveUI 4 settings folder, ready to be used by ActiveUI 5", () => {
    expect(migrateSettingsFolder(legacySettingsFolder)).toMatchSnapshot();
  });

  it("returns an empty 'organization_settings' file and an empty 'users' folder when there is no settings folder", () => {
    expect(migrateSettingsFolder()).toMatchSnapshot();
  });
});
