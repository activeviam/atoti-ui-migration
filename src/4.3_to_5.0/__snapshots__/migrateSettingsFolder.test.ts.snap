// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrateSettingsFolder returns an empty 'organization_settings' file and an empty 'users' folder when there is no settings folder 1`] = `
Object {
  "organization_settings": Object {
    "entry": Object {
      "content": "{}",
      "isDirectory": false,
      "owners": Array [
        "ROLE_CS_ROOT",
      ],
      "readers": Array [
        "ROLE_USER",
      ],
    },
  },
  "users": Object {
    "entry": Object {
      "isDirectory": true,
      "owners": Array [
        "ROLE_USER",
      ],
      "readers": Array [
        "ROLE_USER",
      ],
    },
  },
}
`;

exports[`migrateSettingsFolder returns the folders corresponding to the converted ActiveUI 4 settings folder, ready to be used by ActiveUI 5 1`] = `
Object {
  "organization_settings": Object {
    "entry": Object {
      "content": "{\\"theme\\":\\"light-activeviam\\"}",
      "owners": Array [
        "ROLE_CS_ROOT",
      ],
      "readers": Array [
        "ROLE_USER",
      ],
    },
  },
  "users": Object {
    "children": Object {
      "admin": Object {
        "children": Object {
          "activity": Object {
            "entry": Object {
              "content": "{\\"userFilters\\":[\\"[Geography].[City].[ALL].[AllMember].[Berlin]\\"]}",
              "owners": Array [
                "admin",
              ],
              "readers": Array [
                "admin",
              ],
            },
          },
          "settings": Object {
            "entry": Object {
              "content": "{\\"theme\\":\\"dark-activeviam\\",\\"search.maxResults\\":10,\\"userFilters.areEnabled\\":true}",
              "owners": Array [
                "admin",
              ],
              "readers": Array [
                "admin",
              ],
            },
          },
        },
        "entry": Object {
          "canRead": true,
          "canWrite": true,
          "isDirectory": true,
          "lastEditor": "admin",
          "owners": Array [
            "admin",
          ],
          "readers": Array [
            "admin",
          ],
          "timestamp": 1607879735685,
        },
      },
      "user1": Object {
        "children": Object {
          "activity": Object {
            "entry": Object {
              "content": "{}",
              "owners": Array [
                "user1",
              ],
              "readers": Array [
                "user1",
              ],
            },
          },
          "settings": Object {
            "entry": Object {
              "content": "{\\"userFilters.areEnabled\\":false}",
              "owners": Array [
                "user1",
              ],
              "readers": Array [
                "user1",
              ],
            },
          },
        },
        "entry": Object {
          "canRead": true,
          "canWrite": true,
          "isDirectory": true,
          "lastEditor": "admin",
          "owners": Array [
            "user1",
          ],
          "readers": Array [
            "user1",
          ],
          "timestamp": 1607879735685,
        },
      },
    },
    "entry": Object {
      "isDirectory": true,
      "owners": Array [
        "ROLE_USER",
      ],
      "readers": Array [
        "ROLE_USER",
      ],
    },
  },
}
`;
